src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeGeneration/Symbols/ICodeGenerationTypeSymbol.cs

```diff
@@ -0,0 +1,190 @@
+    string ToDisplayString(NullableFlowState topLevelNullability, SymbolDisplayFormat? format = null);
+
+    /// <summary>
+    /// Converts a symbol to an array of string parts, each of which has a kind. Useful
+    /// for colorizing the display string.
+    /// </summary>
+    /// <param name="topLevelNullability">The top-level nullability to use for formatting.</param>
+    /// <param name="format">Format or null for the default.</param>
+    /// <returns>A read-only array of string parts.</returns>
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(NullableFlowState topLevelNullability, SymbolDisplayFormat? format = null);
+
+    /// <summary>
+    /// Converts a symbol to a string that can be displayed to the user. May be tailored to a
+    /// specific location in the source code.
+    /// </summary>
+    /// <param name="semanticModel">Binding information (for determining names appropriate to
+    /// the context).</param>
+    /// <param name="topLevelNullability">The top-level nullability to use for formatting.</param>
+    /// <param name="position">A position in the source code (context).</param>
+    /// <param name="format">Formatting rules - null implies <see cref="SymbolDisplayFormat.MinimallyQualifiedFormat"/></param>
+    /// <returns>A formatted string that can be displayed to the user.</returns>
+    string ToMinimalDisplayString(
+        SemanticModel semanticModel,
+        NullableFlowState topLevelNullability,
+        int position,
+        SymbolDisplayFormat? format = null);
+
+    /// <summary>
+    /// Convert a symbol to an array of string parts, each of which has a kind. May be tailored
+    /// to a specific location in the source code. Useful for colorizing the display string.
+    /// </summary>
+    /// <param name="semanticModel">Binding information (for determining names appropriate to
+    /// the context).</param>
+    /// <param name="topLevelNullability">The top-level nullability to use for formatting.</param>
+    /// <param name="position">A position in the source code (context).</param>
+    /// <param name="format">Formatting rules - null implies <see cref="SymbolDisplayFormat.MinimallyQualifiedFormat"/></param>
+    /// <returns>A read-only array of string parts.</returns>
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayParts(
+        SemanticModel semanticModel,
+        NullableFlowState topLevelNullability,
+        int position,
+        SymbolDisplayFormat? format = null);
+
+    /// <summary>
+    /// Nullable annotation associated with the type, or <see cref="NullableAnnotation.None"/> if there are none.
+    /// </summary>
+    NullableAnnotation NullableAnnotation { get; }
+
+    /// <summary>
+    /// Returns the same type as this type but with the given nullable annotation.
+    /// </summary>
+    /// <param name="nullableAnnotation">The nullable annotation to use</param>
+    ITypeSymbol WithNullableAnnotation(NullableAnnotation nullableAnnotation);
+}
```

Comment:
it seems like this is going to a be a major maintenance headache.  we now have an entire copy of the ISymbol hierarchy here.  The purpose of the ISYmbol hierarchy was precisely so that the IDE code and compiler code were sharing teh exact same symbol hierarchy.  that's why we moved it down to that layer originally (it started as an IDE-only layer).

Reply:
Yes and no. Not all members from the corresponding `ISymbol` need to be defined here (only ones we want to use). Also, any members which _are_ present are validated for correct signatures.